xarray/tests/test_backends.py

```diff
@@ -2347,6 +2347,48 @@ def setup_and_verify_store(expected=data):
             ):
                 data2.to_zarr(store, region={"x": slice(3)})
 
+    def test_write_region_w_coords(self):
+        data = Dataset(
+            {
+                "u": (
+                    (
+                        "x",
+                        "y",
+                    ),
+                    np.array([[10], [11], [12]]),
+                )
+            },
+            coords={"x": [0, 1, 2], "y": [0]},
+        )
+        data2 = Dataset(
+            {
+                "u": (
+                    (
+                        "x",
+                        "y",
+                    ),
+                    np.array([[13], [14]]),
+                )
+            },
+            coords={"x": [3, 4], "y": [0]},
```

Comment:
I think you'll have to account for non-dimension coordinate variables e.g. `"z": ("x", [3, 4])` so doing something like `set(ds) - set(ds.xindexes)`

Reply:
Okay, I think I understood now. @dcherian, please have a look the updated pull request.